### ISSUE TITLE
Partial_sell not executed properly causing many tries ending up in emergency exit

### ISSUE THREAD

**Issue by the-master006** [freqtrade/freqtrade#7365]
## Describe your environment

  * Operating system: Ubuntu 20.04
  * Python Version: Default from latest stable docker image (3.9? freqtrade 2022.8 - stable ) (`python -V`)
  * CCXT version: Default from latest stable docker image (freqtrade 2022.8 - stable) (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2022.8 stable (`freqtrade -V` or `docker-compose run --rm freqtrade -V` for Freqtrade running in docker)
  
Note: All issues other than enhancement requests will be closed without further comment if the above template is deleted or not filled out.

## Describe the problem:

While dry-running a bot on OKX which is using partial-exit: the moment when a partial exit happens, telegram gives this error, and keeps on repeating it at least 100 times:
![image](https://user-images.githubusercontent.com/65864740/188578290-a70dabc6-22fc-47c1-b3c6-531ff6e1f747.png)
See the logging from freqtrade below:

### Steps to reproduce:

  1. create a strategy with partial sell; see my relevant code below
  2. let it dry-run.
  3. when a partial sell event happens I get this error. It can take 15 repeats of this error or even ~300 times before it stops and apparently the order is executed by an "emergency_exit". Does happen on every partial exit that needs to be executed.
  4. 
  
### Observed Results:

  * What happened: Order gets cancelled on exchange from 10s to 100s of times until it exits by an emergency_exit
  * What did you expect to happen? Partial_exit gets filled and executed for the first part, somewhat later the second partial_exit will be filled and executed and so on up till 3 parial_exits and the last part on ROI or another exit reason

### Relevant code exceptions or logs

Note: Please copy/paste text of the messages, no screenshots of logs please.
```
2022-09-05 15:15:47 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=3, pair=BTC/USDT:USDT, amount=0.02000000, is_short=False, leverage=1.0, open_rate=19752.30000000, open_since=2022-09-05 12:31:03)
2022-09-05 15:15:47 - Aurox_AI_renewed_higher_timefrs - INFO - bladiebla bot loop start
2022-09-05 15:15:48 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=3, pair=BTC/USDT:USDT, amount=0.02000000, is_short=False, leverage=1.0, open_rate=19752.30000000, open_since=2022-09-05 12:31:03)
2022-09-05 15:15:48 - freqtrade.freqtradebot - INFO - Sell order cancelled on exchange for Trade(id=3, pair=BTC/USDT:USDT, amount=0.02000000, is_short=False, leverage=1.0, open_rate=19752.30000000, open_since=2022-09-05 12:31:03).
2022-09-05 15:15:48 - freqtrade.wallets - INFO - Wallets synced.
2022-09-05 15:15:48 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit_cancel, 'trade_id': 3, 'exchange': 'Okx', 'pair': 'BTC/USDT:USDT', 'leverage': 1.0, 'direction': 'Long', 'gain': 'loss', 'limit': 0, 'order_type': 'market', 'amount': 0.0, 'open_rate': 19752.3, 'current_rate': 19886.8, 'profit_amount': -395.1250092, 'profit_ratio': -1.0, 'buy_tag': '', 'enter_tag': '', 'sell_reason': None, 'exit_reason': None, 'open_date': datetime.datetime(2022, 9, 5, 12, 31, 3, 413461), 'close_date': datetime.datetime(2022, 9, 5, 15, 15, 48, 198768, tzinfo=datetime.timezone.utc), 'stake_currency': 'USDT', 'fiat_currency': 'USD', 'reason': 'cancelled on exchange', 'sub_trade': True, 'stake_amount': 395.046}
```

my relevant code:

    def adjust_trade_position(self, trade: Trade, current_time: datetime,
                              current_rate: float, current_profit: float,
                              max_stake: float,
                              current_entry_rate: float, current_exit_rate: float,
                              current_entry_profit: float, current_exit_profit: float,
                              **kwargs) -> float:
        """
        Custom trade adjustment logic, returning the stake amount that a trade should be
        increased or decreased.
        This means extra buy or sell orders with additional fees.
        Only called when `position_adjustment_enable` is set to True.

        """

        if current_profit > (self.sell_0.value/100) and trade.nr_of_successful_exits == 0:
            # Take half of the profit at +0.6%
            return -(trade.stake_amount / 4)

        if current_profit > (self.sell_1.value/100) and trade.nr_of_successful_exits == 1:
            # Take half of the profit at +1%
            return -(trade.stake_amount / 4)

        if current_profit > (self.sell_2.value/100) and trade.nr_of_successful_exits == 2:
            # Take half of the profit at +4%
            return -(trade.stake_amount / 4)

        return None

Also from the strategy:

    # Optional order time in force.
    order_time_in_force = {
        'entry': 'gtc',
        'exit': 'gtc'
    }

My config.json relevant parameters:
```
{
  "max_open_trades": 2,
  "stake_currency": "USDT",
  "dry_run_wallet": 1000,
  "stake_amount": "unlimited",
  "amend_last_stake_amount ": true,
  "position_adjustment_enable": true,
  "tradable_balance_ratio": 0.99,
  "fiat_display_currency": "USD",
  "dry_run": true,
  "available_capital": 990,
  "cancel_open_orders_on_exit": false,
  "trading_mode": "futures",
  "margin_mode": "isolated",
  "unfilledtimeout": {
    "entry": 60,
    "exit": 60,
    "exit_timeout_count": 60,
    "unit": "seconds"
  },
  "entry_pricing": {
    "price_side": "same",
    "use_order_book": true,
    "order_book_top": 1,
    "price_last_balance": 0.0,
    "check_depth_of_market": {
      "enabled": false,
      "bids_to_ask_delta": 1
    }
  },
  "exit_pricing": {
    "price_side": "other",
    "use_order_book": true,
    "order_book_top": 1
  },
  "order_types": {
    "entry": "limit",
    "exit": "market",
    "emergency_exit": "market",
    "force_exit": "market",
    "force_entry": "market",
    "stoploss": "market",
    "stoploss_on_exchange": false,
    "stoploss_on_exchange_interval": 60,
    "stoploss_on_exchange_limit_ratio": 0.99
  },

    "ccxt_async_config": {
      "enableRateLimit": true,
      "rateLimit": 200
    },
```

**Comment by the-master006** [comment]
This behavior might also be the cause for the next error, which happens now for several times already, causing freqtrade to die:
I think this is related somehow:

```
2022-09-06 14:31:45 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': warning, 'status': 'Dry run is enabled. All trades are simulated.'}
2022-09-06 14:31:45 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': startup, 'status': "*Exchange:* `okx`\n*Stake per trade:* `unlimited USDT`\n*Minimum ROI:* `{'0': 0.201, '61': 0.06, '162': 0.039, '315': 0}`\n*Stoploss:* `-0.035`\n*Position adjustment:* `On`\n*Timeframe:* `15m`\n*Strategy:* `Aurox_AI_renewed_higher_timefrs`"}
2022-09-06 14:31:45 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': startup, 'status': "Searching for USDT pairs to buy and sell based on [{'StaticPairList': 'StaticPairList'}]"}
2022-09-06 14:31:46 - freqtrade.persistence.trade_model - INFO - Found open trade: Trade(id=19, pair=ETH/USDT:USDT, amount=0.30000000, is_short=True, leverage=1.0, open_rate=1668.81000000, open_since=2022-09-06 13:01:41)
2022-09-06 14:31:47 - Aurox_AI_renewed_higher_timefrs - INFO - bladiebla bot loop start
2022-09-06 14:31:48 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=19, pair=ETH/USDT:USDT, amount=0.30000000, is_short=True, leverage=1.0, open_rate=1668.81000000, open_since=2022-09-06 13:01:41)
2022-09-06 14:31:48 - freqtrade.freqtradebot - INFO - Buy order cancelled on exchange for Trade(id=19, pair=ETH/USDT:USDT, amount=0.30000000, is_short=True, leverage=1.0, open_rate=1668.81000000, open_since=2022-09-06 13:01:41).
2022-09-06 14:31:48 - freqtrade.wallets - INFO - Wallets synced.
2022-09-06 14:31:48 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit_cancel, 'trade_id': 19, 'exchange': 'Okx', 'pair': 'ETH/USDT:USDT', 'leverage': 1.0, 'direction': 'Short', 'gain': 'loss', 'limit': 0, 'order_type': 'market', 'amount': 0.0, 'open_rate': 1668.81, 'current_rate': 1650.18, 'profit_amount': 500.3926785, 'profit_ratio': 0.0, 'buy_tag': '', 'enter_tag': '', 'sell_reason': None, 'exit_reason': None, 'open_date': datetime.datetime(2022, 9, 6, 13, 1, 41, 124661), 'close_date': datetime.datetime(2022, 9, 6, 14, 31, 48, 196265, tzinfo=datetime.timezone.utc), 'stake_currency': 'USDT', 'fiat_currency': 'USD', 'reason': 'cancelled on exchange', 'sub_trade': True, 'stake_amount': 500.6430000000001}
2022-09-06 14:31:49 - freqtrade.wallets - INFO - Wallets synced.
2022-09-06 14:31:49 - freqtrade.commands.trade_commands - ERROR - float division by zero
2022-09-06 14:31:49 - freqtrade.commands.trade_commands - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/freqtrade/freqtrade/commands/trade_commands.py", line 19, in start_trading
    worker.run()
  File "/freqtrade/freqtrade/worker.py", line 74, in run
    state = self._worker(old_state=state)
  File "/freqtrade/freqtrade/worker.py", line 114, in _worker
    self._throttle(func=self._process_running, throttle_secs=self._throttle_secs)
  File "/freqtrade/freqtrade/worker.py", line 139, in _throttle
    result = func(*args, **kwargs)
  File "/freqtrade/freqtrade/worker.py", line 152, in _process_running
    self.freqtrade.process()
  File "/freqtrade/freqtrade/freqtradebot.py", line 211, in process
    self.process_open_trade_positions()
  File "/freqtrade/freqtrade/freqtradebot.py", line 531, in process_open_trade_positions
    self.check_and_call_adjust_trade_position(trade)
  File "/freqtrade/freqtrade/freqtradebot.py", line 595, in check_and_call_adjust_trade_position
    self.execute_trade_exit(trade, current_exit_rate, exit_check=ExitCheckTuple(
  File "/freqtrade/freqtrade/freqtradebot.py", line 1560, in execute_trade_exit
    self._notify_exit(trade, order_type, sub_trade=bool(sub_trade_amt), order=order_obj)
  File "/freqtrade/freqtrade/freqtradebot.py", line 1583, in _notify_exit
    profit_ratio = trade.calc_profit_ratio(profit_rate, amount, trade.open_rate)
  File "/freqtrade/freqtrade/persistence/trade_model.py", line 832, in calc_profit_ratio
    profit_ratio = (1 - (close_trade_value / open_trade_value)) * leverage
ZeroDivisionError: float division by zero
2022-09-06 14:31:49 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2022-09-06 14:31:49 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2022-09-06 14:31:49 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2022-09-06 14:31:49 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': warning, 'status': "1 open trades active.\n\nHandle these trades manually on OKX, or '/start' the bot again and use '/stopentry' to handle open trades gracefully. \nNote: Trades are simulated (dry run)."}
```

**Comment by xmatthias** [comment]
Please use `freqtrade show-trades --db-url sqlite:///<trades.sqlite> --print-json --trade-ids 3` to get the trade output - so we can see what the bot actually tried to do.

we'll also need more logs from the exit time. it only shows when the order is cancelled - but is lacking information on the order placement - and a few lines before that.


based on this i hope we can see some more. 
i can't however recreate this for now - the trades / orders complete just fine.

The error is also quite strange - as you're using market exit orders - which will fill immediately.

**Comment by the-master006** [comment]
Hereby the output from that command. I am not sure how I can get more logs from the exit. If this same issue happens again (which is with every positive partial exit, I will gather it and post it.

```
2022-09-06 17:49:45,862 - freqtrade.configuration.load_config - INFO - Using config: user_data/config.json ...
2022-09-06 17:49:45,868 - freqtrade.loggers - INFO - Verbosity set to 0
2022-09-06 17:49:45,871 - freqtrade.configuration.configuration - INFO - Parameter --db-url detected ...
2022-09-06 17:49:45,925 - freqtrade.configuration.configuration - INFO - Using user-data directory: /freqtrade/user_data ...
2022-09-06 17:49:45,926 - freqtrade.configuration.configuration - INFO - Using data directory: /freqtrade/user_data/data/okx ...
2022-09-06 17:49:45,927 - freqtrade.configuration.configuration - INFO - Parameter --print-json detected ...
2022-09-06 17:49:45,928 - freqtrade.configuration.configuration - INFO - Filtering on trade_ids: ['3']
2022-09-06 17:49:45,929 - freqtrade.configuration.check_exchange - INFO - Checking exchange...
2022-09-06 17:49:45,945 - freqtrade.configuration.check_exchange - INFO - Exchange "okx" is officially supported by the Freqtrade development team.
2022-09-06 17:49:45,947 - freqtrade.configuration.configuration - INFO - Using pairlist from configuration.
2022-09-06 17:49:45,948 - freqtrade.configuration.config_validation - INFO - Validating configuration ...
2022-09-06 17:49:45,954 - freqtrade.commands.list_commands - INFO - Using DB: "sqlite:////freqtrade/user_data/tradesv35.sqlite"
2022-09-06 17:49:46,130 - freqtrade.commands.list_commands - INFO - Printing 1 Trades:
[
    {
        "trade_id": 3,
        "pair": "BTC/USDT:USDT",
        "base_currency": "BTC",
        "quote_currency": "USDT",
        "is_open": false,
        "exchange": "okx",
        "amount": 0.02,
        "amount_requested": 0.02501525,
        "stake_amount": 395.046,
        "strategy": "Aurox_AI_renewed_higher_timefrs",
        "buy_tag": "",
        "enter_tag": "",
        "timeframe": 15,
        "fee_open": 0.0002,
        "fee_open_cost": 0.0790092,
        "fee_open_currency": "USDT",
        "fee_close": 0.0005,
        "fee_close_cost": 0.198853,
        "fee_close_currency": "USDT",
        "open_date": "2022-09-05 12:31:03",
        "open_timestamp": 1662381063413,
        "open_rate": 19752.3,
        "open_rate_requested": 19752.3,
        "open_trade_value": 395.1250092,
        "close_date": "2022-09-05 15:15:51",
        "close_timestamp": 1662390951198,
        "realized_profit": 2.3821378,
        "close_rate": 19885.3,
        "close_rate_requested": 19885.3,
        "close_profit": 0.006028820612552611,
        "close_profit_pct": 0.6,
        "close_profit_abs": 2.3821378,
        "trade_duration_s": 9887,
        "trade_duration": 164,
        "profit_ratio": 0.006028820612552611,
        "profit_pct": 0.6,
        "profit_abs": 2.3821378,
        "sell_reason": "force_exit",
        "exit_reason": "force_exit",
        "exit_order_status": "closed",
        "stop_loss_abs": 19061.0,
        "stop_loss_ratio": -0.035,
        "stop_loss_pct": -3.5000000000000004,
        "stoploss_order_id": null,
        "stoploss_last_update": "2022-09-05 12:31:07",
        "stoploss_last_update_timestamp": 1662381067338,
        "initial_stop_loss_abs": 19061.0,
        "initial_stop_loss_ratio": -0.035,
        "initial_stop_loss_pct": -3.5000000000000004,
        "min_rate": 19711.2,
        "max_rate": 19916.1,
        "leverage": 1.0,
        "interest_rate": 0.0,
        "liquidation_price": null,
        "is_short": false,
        "trading_mode": "futures",
        "funding_fees": 0.0,
        "open_order_id": null,
        "orders": [
            {
                "amount": 0.02,
                "safe_price": 19752.3,
                "ft_order_side": "buy",
                "order_filled_timestamp": 1662381067320,
                "ft_is_entry": true,
                "pair": "BTC/USDT:USDT",
                "order_id": "dry_run_buy_1662381063.343616",
                "status": "closed",
                "average": 19752.3,
                "cost": 395.046,
                "filled": 0.02,
                "is_open": false,
                "order_date": "2022-09-05 12:31:03",
                "order_timestamp": 1662381063000,
                "order_filled_date": "2022-09-05 12:31:07",
                "order_type": "limit",
                "price": 19752.3,
                "remaining": 0.0
            },
            {
                "amount": 0.02,
                "safe_price": 19885.3,
                "ft_order_side": "sell",
                "order_filled_timestamp": 1662390951153,
                "ft_is_entry": false,
                "pair": "BTC/USDT:USDT",
                "order_id": "dry_run_sell_1662390950.799516",
                "status": "closed",
                "average": 19885.3,
                "cost": 397.706,
                "filled": 0.02,
                "is_open": false,
                "order_date": "2022-09-05 15:15:50",
                "order_timestamp": 1662390950000,
                "order_filled_date": "2022-09-05 15:15:51",
                "order_type": "market",
                "price": 19885.3,
                "remaining": 0.0
            }
        ]
    }
]
```

**Comment by xmatthias** [comment]
hmm ... that's ... not the information i was hoping for ... but it makes sense to not include canceled orders in the output in general ... 

would you mind also posting the output of the following from the database (the below assumes sqlite3 - but you can obviously use any db editor ...)

```
sqlite> .headers on
sqlite> select * from orders where trade_id = 3;
```

**Comment by the-master006** [comment]
hmm sorry, but I don't have a clue how to do this, I am running the docker stable image, how can I access the sqlite database then? Thanks

**Comment by xmatthias** [comment]
you can run `docker-compose  run --entrypoint bash freqtrade` (assuming you didn't change the service name).

from there - run `sqlite3 user_data/trades.sqlite` (or whatever your file is called) - and execute the 2 commands above (everything after the >).

**Comment by the-master006** [comment]
[ft_trade_id 3.txt](https://github.com/freqtrade/freqtrade/files/9499865/ft_trade_id.3.txt)

Put it in a file: it are quite a lot of retries

**Comment by xmatthias** [comment]
i feared this much - all exit orders are with an amount of 0.0 - which count as "not filled" even though they are somehow filled (not really).

---

i think i now know what the problem is - your exits are < minimum tradable (which is currently around 284.55 for BTC/USDT:USDT on okx with leverage 1).

Futures pairs (on okx in particular) have pretty high minimum requirements - especially some high volume pairs like ETH and BTC.
your logic is in an unfortunate position to end up "right to the problem area" with the first partial sell - i've gotten most sell attempts at "around 284.4465" - which is barely below the limit - so it's in theory possible to jump over that limit for a moment.

As we're however also dealing with contracts (a multiplier, basically) - we'll also need to "round" the resulting amount (the minimum tradable is 0.01 BTC on okx futures) - so you'll only be able to sell half of the position (where we're now again WAY below the minimum tradable amount of 284).

The check for this currently doesn't do this "rounding" - which is why you're able to get past that and end up with orders with an amount of 0.0 (after rounding 0.005 BTC to "min tradable, it's 0.0).

That'll hopefully be an easy fix - the result will be that all partial exits for this strategy (at least in this configuration) will be rejected as being below the minimum amount - at least for BTC (which has a pretty high minimum amount).

**Comment by the-master006** [comment]
Ok thanks xmatthias! So if I increase my stakes then this could be a workaround?

**Comment by xmatthias** [comment]
yes - increasing the stakes should work - probably by 10x so you actually get to have 4 exits :laughing: 

Unfortunately, it's not an limitation imposed by freqtrade - you'll notice that the okx website won't allow you to trade < 0.01 BTC on the futures market, either.

You should currently also have log messages around `Remaining amount of 284.5405 would be too small.` in your logs - which suggests rejected partial sells (i'll clarify the message a bit here).

**Comment by xmatthias** [comment]
fixed with 4d69df08dde96d067d9daebbf2db3a61a24dd44a
